src/library/scala/collection/SetLike.scala

```diff
@@ -150,7 +150,35 @@ self =>
    *  @param elems     the collection containing the elements to add.
    *  @return a new $coll with the given elements added, omitting duplicates.
    */
-  def ++ (elems: GenTraversableOnce[A]): This = (repr /: elems.seq)(_ + _)
+  def ++ (elems: GenTraversableOnce[A]): This = {
+    import immutable.HashSet
+    //in 2.14 this should be moved to the appropriate place - HashSet and EmptySet.
+    //we can't break binary comparability before then
+    this match {
+      case _ if this eq immutable.Set.empty.asInstanceOf[AnyRef] =>
+        import immutable.Set.{Set1, Set2, Set3, Set4}
+        elems match {
```

Comment:
Shouldn't the simple version with the cast work for any `elems.isInstanceOf[immutable.Set[_]]`?

Reply:
Its just blindly replacing with `elems` so this can only work with empty sets, and empty sets where we know that elems is of a type that would naturally grow to be a set1-4 of HashSet
e.g. if this eq SortedSet.empty, we would expect to result in a SortedSet

... Unless I  have missed your point of course

Reply:
Ah, right, you wouldn't want to keep the type of more specialized sets. They have to be copied to the default set implementation.